### PR TITLE
fix: improve loadServiceInstanceAndGrantAccess query (#3193)

### PR DESCRIPTION
**Pull request by hervyt**
# Context
> Optim the serviceInstanceById Query

## How to test
- You should see no regression

## Related issues

- Related to #3193 

# Checklist

## Quality

- [X] I consider this work complete and ready for review
- [X] I tested all features and edge cases
- [X] I refactored code where necessary to improve quality
- [X] I made sure my code meets development guidelines
- [X] I performed a self-review of my code
- [ ] I added the `skip-feature-env` label if a feature environment is not needed for this PR

## Testing

- [ ] I wrote test cases for the relevant use cases (unit, integration or e2e tests)
- [ ] (Bug fixes only) I added tests that reproduce and verify the fix

### What tests have been made

- [X] Unit tests
- [X] Integration tests
- [ ] E2E tests
- [X] Local manual tests

### CODE REVIEW

File: apps/backend/src/migrations/20260826122300_remove_non_filigran_subscriptions.js [hunk 1]
```diff
@@ -0,0 +1,17 @@
+/**
+ * @param { import("knex").Knex } knex
+ * @returns { Promise<void> }
+ */
+export async function up(knex) {
+  await knex('Subscription')
+    .whereNot('organization_id', 'ba091095-418f-4b4f-b150-6c9295e232c4')
+    .delete();
+}
```
Comment (jpkha) [inline apps/backend/src/migrations/20260826122300_remove_non_filigran_subscriptions.js]: Are we sure about this ? Only Filigran needs subscription ? What about the trials ?

Comment (hervyt) [reply apps/backend/src/migrations/20260826122300_remove_non_filigran_subscriptions.js]: OUhhhhhh 
Did not think about subscription's trials 😱

Comment (hervyt) [reply apps/backend/src/migrations/20260826122300_remove_non_filigran_subscriptions.js]: Fixed

File: apps/backend/src/modules/security-management/service-capability/service-capability.app.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  userServiceCapabilities: UserServiceCapability[];
+  subscription_id: SubscriptionId | null;
+};
+
+export const ServiceCapabilityApp = {
+  loadServiceCapabilitiesByServiceId: async (
+    serviceInstanceId: ServiceInstanceId
+  ): Promise<UserServiceCapabilitiesResponse> => {
+    const user = requestContext.requireUser();
+    const subscriptionPromise = user.selected_organization_id
```
Comment (jpkha) [inline apps/backend/src/modules/security-management/service-capability/service-capability.app.ts]: is it possible to have an user with a falsy selected_organization_id ?

File: apps/backend/src/modules/security-management/service-capability/service-capability.app.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+      ? await SubscriptionDomain.loadSubscriptionBy({
+          service_instance_id: serviceInstanceId,
+          organization_id: user.selected_organization_id,
+        })
+      : Promise.resolve(undefined);
```
Comment (jpkha) [inline apps/backend/src/modules/security-management/service-capability/service-capability.app.ts]: it seems weird to me to return a Promise and the other side a resolve of the Promise.
We should return a Promise or else the result of the promise, I think

File: apps/frontend/src/components/service/components/header/ServiceListHeaderButtons.tsx [hunk 4]
```diff
@@ -41,11 +41,11 @@ const ServiceListHeaderButtons = ({}) => {
 
   return (
     <div className="flex gap-s">
-      {(canManageService || isBypass) && currentUserSubscriptionId && (
+      {(canManageService || isAdminOrga || isBypass) && subscriptionId && (
```
Comment (jpkha) [inline apps/frontend/src/components/service/components/header/ServiceListHeaderButtons.tsx]: Why do you add `isAdminOrga`, does `canManageService` already manage it ?

File: apps/frontend/app/(application)/app/(user)/service/opencti_custom_dashboards/[serviceInstanceId]/page.tsx [hunk 5]
```diff
@@ -42,7 +42,7 @@ const Page = async ({ params }: ServiceCustomDashboardsPageProps) => {
           <PageLoader
             serviceInstance={
               response.data
-                .serviceInstanceByIdAndGrantAccess as unknown as serviceInstance_fragment$data
+                .serviceInstanceById as unknown as serviceInstance_fragment$data
```
Comment (jpkha) [inline apps/frontend/app/(application)/app/(user)/service/opencti_custom_dashboards/[serviceInstanceId]/page.tsx:45]: do we still need as unkown ?

File: apps/backend/src/modules/security-management/service-capability/service-capability.app.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+      ),
+      subscriptionPromise,
+    ]);
+
+    const userServiceCapabilities = serviceCapabilities.filter(
+      (
+        serviceCapability
+      ): serviceCapability is UserServiceCapabilityModel & {
+        user_service_id: string;
+      } => serviceCapability.user_service_id !== null
```
Comment (Kroustille) [inline apps/backend/src/modules/security-management/service-capability/service-capability.app.ts]: There's no need to apply this filter because `ServiceCapabilityDomain.loadServiceCapabilitiesByServiceId` returns a non nullable `user_service_id`

File: apps/backend/src/modules/security-management/service-capability/service-capability.app.ts [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+  subscription_id: SubscriptionId | null;
+};
+
+export const ServiceCapabilityApp = {
+  loadServiceCapabilitiesByServiceId: async (
+    serviceInstanceId: ServiceInstanceId
+  ): Promise<UserServiceCapabilitiesResponse> => {
+    const user = requestContext.requireUser();
+    const subscriptionPromise = user.selected_organization_id
+      ? await SubscriptionDomain.loadSubscriptionBy({
```
Comment (Kroustille) [inline apps/backend/src/modules/security-management/service-capability/service-capability.app.ts]: The subscription is already joined and returned in with `ServiceCapabilityDomain.loadServiceCapabilitiesByServiceId` query. I'm not sure it's required to prefetch this subscription 🤔

File: apps/backend/src/modules/security-management/service-capability/service-capability.app.ts (hunk 2 above)
Comment (hervyt) [reply apps/backend/src/modules/security-management/service-capability/service-capability.app.ts]: Nope, can't be null | undefined

File: apps/frontend/src/components/service/components/header/ServiceListHeaderButtons.tsx (hunk 4 above)
Comment (hervyt) [reply apps/frontend/src/components/service/components/header/ServiceListHeaderButtons.tsx]: I rename the var, the canManageService is only the service capa

File: apps/frontend/src/hooks/use-service-capability.tsx [hunk 8]
```diff
@@ -1,21 +1,78 @@
...
   capability: ServiceRestriction,
   serviceInstance?: serviceInstance_fragment$data
-) => {
+): boolean;
+function useServiceCapability(
+  capability: ServiceRestriction,
+  serviceInstance: serviceInstance_fragment$data | undefined,
+  options: { withSubscriptionId: true }
+): UseServiceCapabilityWithSubscriptionId;
+function useServiceCapability(
```
Comment (Kroustille) [inline apps/frontend/src/hooks/use-service-capability.tsx]: I find it confusing to have the same hook returning either a `boolean` or `UseServiceCapabilityWithSubscriptionId` depending on the given options 🤔

File: apps/backend/src/modules/security-management/service-capability/service-capability.resolver.ts [hunk 9]
```diff
@@ -1,9 +1,23 @@
...
   ServiceCapabilityId:
     createRelayIdScalar<ServiceCapabilityId>('Service_Capability'),
+  Query: {
+    userServiceCapabilities: async (_, { service_instance_id }) => {
+      try {
+        return await ServiceCapabilityApp.loadServiceCapabilitiesByServiceId(
+          service_instance_id
+        );
+      } catch (error) {
+        throw mapToGraphQLError(error, UnknownErrorCode.AddCapabilitiesError);
```
Comment (Kroustille) [inline apps/backend/src/modules/security-management/service-capability/service-capability.resolver.ts:18]: The error does not seem to fit with the query

File: apps/backend/src/modules/security-management/service-capability/service-capability.domain.ts [hunk 10]
```diff
@@ -20,6 +26,91 @@ export const ServiceCapabilityDomain = {
...
+        '=',
+        'Subscription_Capability.id'
+      )
+      .leftJoin(
+        'Service_Capability',
+        'Subscription_Capability.service_capability_id',
+        '=',
+        'Service_Capability.id'
+      )
+      .tap(restrictSubscriptionToUserOrganization)
```
Comment (Kroustille) [inline apps/backend/src/modules/security-management/service-capability/service-capability.domain.ts]: This is a minor performance issue: using tap here adds a new join to subscription but we already have the subscription table joined in this query.
This tap can be replaced by `.andWhere('Subscription.organization_id', user.selected_organization_id)`

File: apps/backend/src/modules/security-management/service-capability/service-capability.app.ts [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+          service_instance_id: serviceInstanceId,
+          organization_id: user.selected_organization_id,
+        })
+      : Promise.resolve(undefined);
+
+    const [serviceCapabilities, subscription] = await Promise.all([
```
Comment (Kroustille) [inline apps/backend/src/modules/security-management/service-capability/service-capability.app.ts:23]: This `Promise.all` is not useful because `subscriptionPromise` is already an awaited result. Requests still runs sequentially.

File: apps/frontend/app/(application)/app/(user)/service/opencti_custom_dashboards/[serviceInstanceId]/page.tsx (hunk 5 above)
Comment (hervyt) [reply apps/frontend/app/(application)/app/(user)/service/opencti_custom_dashboards/[serviceInstanceId]/page.tsx:45]: Yes, otherwise we have a type error; maybe we can handle all these as unknown in a dedicated PR everywhere ?

File: apps/backend/src/modules/security-management/service-capability/service-capability.app.ts (hunk 7 above)
Comment (hervyt) [reply apps/backend/src/modules/security-management/service-capability/service-capability.app.ts]: It is in the case the user is only AdminOrga and does not have serviceCapabilities + loadServiceCapabilitiesByServiceId should not return this subscription_id anymore 🧑‍🎤  I remove it 📦

File: apps/backend/src/modules/security-management/service-capability/service-capability.app.ts (hunk 3 above)
Comment (hervyt) [reply apps/backend/src/modules/security-management/service-capability/service-capability.app.ts]: Now it's 2 promises :)

File: apps/backend/src/modules/security-management/service-capability/service-capability.app.ts (hunk 11 above)
Comment (hervyt) [reply apps/backend/src/modules/security-management/service-capability/service-capability.app.ts:23]: Promise.all is still useful here: subscriptionPromise is not awaited before Promise.all, so both async calls start first and are then awaited together.
 They run concurrently, not sequentially

Comment (Kroustille) [reply apps/backend/src/modules/security-management/service-capability/service-capability.app.ts:23]: Haha I saw an `await` when I posted the comment 😛 Looks good now 👍

File: apps/backend/src/modules/security-management/service-capability/service-capability.app.ts (hunk 2 above)
Comment (Kroustille) [reply apps/backend/src/modules/security-management/service-capability/service-capability.app.ts]: If it can't be `null | undefined`, it's not required to use the ternary to have `subscriptionPromise`

File: apps/backend/src/modules/security-management/service-capability/service-capability.app.test.ts [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+describe('serviceCapability app', () => {
+  const serviceInstanceId = SERVICES.INSTANCES.EPIC.ID;
+  const userId = TEST_ORGANIZATIONS.SECOND_ORGANIZATION.USERS.SIMPLE.ID;
+  const selectedOrganizationId = TEST_ORGANIZATIONS.SECOND_ORGANIZATION.ID;
+
+  afterEach(() => {
+    vi.restoreAllMocks();
+  });
+
+  it('should propagate domain error when loading capabilities fails', async () => {
```
Comment (Kroustille) [inline apps/backend/src/modules/security-management/service-capability/service-capability.app.test.ts:81]: This tests should assert the different use cases in the app fucntion and by calling DB (integration test)

**Review by Kroustille — APPROVED** [review]
